### PR TITLE
planner, statistics: add more debug trace

### PR DESCRIPTION
**Pull request by time-and-fate**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/43069

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/debugtrace.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	BinaryParamsInfo []binaryParamInfo
+	UseCursor        bool
+}
+
+type binaryParamInfo struct {
+	Type  string
+	Value string
+}
+
+func (info *binaryParamInfo) MarshalSON() ([]byte, error) {
```
Comment (xuyifangreeneyes) [inline planner/core/debugtrace.go:54]:
```suggestion
func (info *binaryParamInfo) MarshalJSON() ([]byte, error) {
```

File: statistics/index.go [hunk 2]
```diff
@@ -399,7 +467,7 @@ func (idx *Index) expBackoffEstimation(sctx sessionctx.Context, coll *HistColl,
 }
 
 func (idx *Index) checkStats() {
-	if idx.IsFullLoad() {
+	if idx.IsFullLoad() || idx.PhysicalID <= 0 {
```
Comment (xuyifangreeneyes) [inline statistics/index.go:470]: When does `idx.PhysicalID <= 0` happen?

**Review by xuyifangreeneyes — APPROVED** [review]
rest LGTM

File: statistics/index.go (hunk 2 above)
Comment (time-and-fate) [reply statistics/index.go:471]: Added a comment here.

**Review by winoros — APPROVED** [review]

File: planner/core/debugtrace.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+	}
+	root.AppendStepToCurrentContext(traceInfo)
+}
+
+/*
+ Below is debug trace for AccessPath.
+*/
+
+type accessPathForDebugTrace struct {
+	IndexName        string `json:",omitempty"`
```
[the remark this answers is not in the document]
Comment (time-and-fate) [reply planner/core/debugtrace.go:204]: It has. This tag means using its default name, which is `IndexName`.
You can confirm this in the test result, which is in `server/testdata/optimizer_suite_out.json`.

**Pull request merged**